### PR TITLE
Feature/rud control init

### PR DESCRIPTION
**Pull request by jazreen**
### Description

- Resolves #_issue_

added controls.c and h, containing basic pid logic

### Verification

- [ ] 

### Resources

- 
https://ubcsailbot.atlassian.net/wiki/spaces/prjt22/pages/1890025971/PID+Control+Plan

### CODE REVIEW

File: projects/rudder-controller/project/Core/Src/controls.c [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+	//PWM overvoltage protection***
+	if (PWM_value > ADC_MAX) {
+		PWM_value = ADC_MAX;
+	}
+
+	else if (PWM_value < ADC_MIN) {
+		PWM_value = ADC_MIN;
+	}
+
```
Comment (MoizLokh) [inline projects/rudder-controller/project/Core/Src/controls.c:119]: Better to have this as a seperate function - `compute_pwm`? useful if we plan to modify computation in the future.

File: projects/rudder-controller/project/Core/Src/controls.c [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+	else if (adjusted_error <= ERROR_MIN) {
+		adjusted_error = ERROR_MIN;
+	}
+
+	sum_errors += adjusted_error; //accumulated error over time, used for Pi term calculation
+    I_term = KI * sum_errors; //calculate i term
+
+	//calculate d term from the differential between previous error and current error
+    D_term = KD * (current_error - prev_error);
+    prev_error = current_error;
```
Comment (MoizLokh) [inline projects/rudder-controller/project/Core/Src/controls.c:105]: Fix indentation

File: projects/rudder-controller/project/Core/Src/controls.c [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+	}
+	//COME BACK TO THIS- HAVE TO CONSIDER THE CASE WHERE WE TRIGGER PID
+	*/
+
+	//adjusting error values
+	if (error > HALF_ROTATION) {adjusted_error = error - FULL_ROTATION;} //so it now moves port not stbd, though error is +
+	if (error < -HALF_ROTATION) {adjusted_error = error + FULL_ROTATION;} //so it now moves stbd nor port, though error is -
+
+
+	current_error = abs(adjusted_error); //does it have to be the absolute?
```
Comment (MoizLokh) [inline projects/rudder-controller/project/Core/Src/controls.c:84]: Why are we using abs error to calculate D_term? Do we need abs value?

File: projects/rudder-controller/project/Core/Src/controls.c [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+	P_term = KP * current_error; //calculate p term
+
+	//make sure we dont have integral windup (for when we have sudden change in current/desired heading)
+	//since we'll be using error polarity to determine which way to move, does it make sense to not change error var
+	//directly? we can use a different param just for calculations? i changed eror to adjusted_error here
```
Comment (MoizLokh) [inline projects/rudder-controller/project/Core/Src/controls.c:90]: Are these relavent comments? if not remove any irrelavent comments in the code

File: projects/rudder-controller/project/Core/Src/controls.c [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+		HAL_Delay(DELAY_TIME);
+	}
+}
+
+void run_pid(void){
+	//change to double
+	static int sum_errors = 0, prev_error = 0;
+	int P_term, I_term, D_term;
+	int error, adjusted_error, current_error;
+	int PWM_value, PWM_temp = 100; //gotta change this
```
Comment (MoizLokh) [inline projects/rudder-controller/project/Core/Src/controls.c:56]: More descriptive name for PWM_temp is needed. 100 is a magic constant. Better to use #define

File: projects/rudder-controller/project/Core/Src/controls.c [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+
+	while (1){
+		//add watchdog timer here
+		run_pid();
+		HAL_Delay(DELAY_TIME);
+	}
+}
+
+void run_pid(void){
+	//change to double
```
Comment (MoizLokh) [inline projects/rudder-controller/project/Core/Src/controls.c:52]: Are we changing to double?

File: projects/rudder-controller/project/Core/Src/controls.c [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+	else if (adjusted_error < 0) {
+		motor_move_port();
+	}
+
+
+	//sends the pwm to motor
+	motor_set_duty_cycle(PWM_value);
+
+	PWM_temp = PWM_value;
+
```
Comment (MoizLokh) [inline projects/rudder-controller/project/Core/Src/controls.c:135]: Remove this space too to clean up code

**Review by MoizLokh — CHANGES_REQUESTED** [review]

File: projects/rudder-controller/project/Core/Src/controls.c (hunk 5 above)
Comment (jazreen) [reply projects/rudder-controller/project/Core/Src/controls.c:56]: Im not a 100% sure about what the PWM_temp value should be initially (or how to figure it out). Will change it after talking to Daniel.

**Review by MoizLokh — APPROVED** [review]
Changes look good

**Pull request merged**